File: flaskrfid/routes/purchases.py
from flask import Blueprint, request, jsonify, current_app
import structlog

logger = structlog.get_logger()
purchases_bp = Blueprint('purchases', __name__)

@purchases_bp.route('/foods', methods=['GET'])
def get_foods():
    try:
        purchase_service = current_app.purchase_service
        
        if not purchase_service:
            return jsonify({
                'success': False,
                'error': 'Purchase service not available'
            }), 503
        
        foods = purchase_service.get_available_foods()
        
        return jsonify({
            'success': True,
            'data': foods
        })
    except Exception as e:
        logger.error("Get available foods error", error=str(e))
        
        return jsonify({
            'success': False,
            'error': 'Failed to get available foods'
        }), 500

@purchases_bp.route('/complete', methods=['POST'])
def complete_purchase():
    try:
        purchase_service = current_app.purchase_service
        
        if not purchase_service:
            return jsonify({
                'success': False,
                'error': 'Purchase service not available'
            }), 503
        
        data = request.get_json()
        
        # Validate purchase data
        if not data.get('user_id') or not data.get('items') or not isinstance(data.get('items'), list) or len(data.get('items')) == 0:
            return jsonify({
                'success': False,
                'error': 'Invalid purchase data'
            }), 400
        
        # Validate payment method
        if data.get('payment_method') not in ['cash', 'monthly_billing']:
            return jsonify({
                'success': False,
                'error': 'Invalid payment method'
            }), 400
        
        # Validate items
        purchase_service.validate_purchase_items(data['items'])
        
        # Calculate and verify total
        calculated_total = purchase_service.calculate_total(data['items'])
        if abs(calculated_total - data.get('total_amount', 0)) > 0.01:
            return jsonify({
                'success': False,
                'error': 'Total amount mismatch'
            }), 400
        
        # For cash payments, validate paid amount
        if data.get('payment_method') == 'cash':
            if not data.get('paid_amount') or data.get('paid_amount') < data.get('total_amount', 0):
                return jsonify({
                    'success': False,
                    'error': 'Insufficient cash amount'
                }), 400
        
        result = purchase_service.complete_purchase(data)
        
        logger.info("Purchase Completed", 
                   purchase_id=result['purchase']['id'],
                   cafeteria=current_app.config.get('CAFETERIA_NAME'),
                   station=current_app.config.get('STATION_ID'),
                   payment_method=data.get('payment_method'))
        
        return jsonify(result)
    except Exception as e:
        logger.error("Complete purchase error", error=str(e))
        
        return jsonify({
            'success': False,
            'error': str(e) or 'Failed to complete purchase'
        }), 400

@purchases_bp.route('/user/<user_id>', methods=['GET'])
def get_user_purchases(user_id):
    try:
        limit = int(request.args.get('limit', 10))
        
        purchase_service = current_app.purchase_service
        
        if not purchase_service:
            return jsonify({
                'success': False,
                'error': 'Purchase service not available'
            }), 503
        
        purchases = purchase_service.get_user_purchases(user_id, limit)
        
        return jsonify({
            'success': True,
            'data': purchases
        })
    except Exception as e:
        logger.error("Get user purchases error", error=str(e))
        
        return jsonify({
            'success': False,
            'error': 'Failed to get user purchases'
        }), 500

File: routes/purchases.js
const express = require('express');
const router = express.Router();
const logger = require('../utils/logger');

// Get available foods for purchase
router.get('/foods', async (req, res) => {
  try {
    const language = req.language || 'en';
    const purchaseService = req.app.locals.services.purchase;
    
    if (!purchaseService) {
      return res.status(503).json({
        success: false,
        error: req.t('systemError')
      });
    }

    const foods = await purchaseService.getAvailableFoods();
    
    res.json({
      success: true,
      data: foods
    });
  } catch (error) {
    logger.error('Get available foods error:', error);
    
    res.status(500).json({
      success: false,
      error: req.t('systemError')
    });
  }
});

// Complete purchase
router.post('/complete', async (req, res) => {
  try {
    const purchaseService = req.app.locals.services.purchase;
    
    if (!purchaseService) {
      return res.status(503).json({
        success: false,
        error: 'Purchase service not available'
      });
    }

    // Validate purchase data
    const { userId, items, totalAmount, paymentMethod, paidAmount } = req.body;
    
    if (!userId || !items || !Array.isArray(items) || items.length === 0) {
      return res.status(400).json({
        success: false,
        error: 'Invalid purchase data'
      });
    }

    // Validate payment method
    if (!['cash', 'monthly_billing'].includes(paymentMethod)) {
      return res.status(400).json({
        success: false,
        error: 'Invalid payment method'
      });
    }

    // Validate items
    await purchaseService.validatePurchaseItems(items);

    // Calculate and verify total
    const calculatedTotal = purchaseService.calculateTotal(items);
    if (Math.abs(calculatedTotal - totalAmount) > 0.01) {
      return res.status(400).json({
        success: false,
        error: 'Total amount mismatch'
      });
    }

    // For cash payments, validate paid amount
    if (paymentMethod === 'cash') {
      if (!paidAmount || paidAmount < totalAmount) {
        return res.status(400).json({
          success: false,
          error: 'Insufficient cash amount'
        });
      }
    }

    const result = await purchaseService.completePurchase(req.body);
    
    logger.logPurchaseEvent('Purchase Completed', result.purchase, {
      cafeteria: process.env.CAFETERIA_NAME,
      station: process.env.STATION_ID,
      paymentMethod: paymentMethod
    });

    res.json(result);
  } catch (error) {
    logger.error('Complete purchase error:', error);
    
    res.status(400).json({
      success: false,
      error: error.message || 'Failed to complete purchase'
    });
  }
});

// Get user purchase history
router.get('/user/:userId', async (req, res) => {
  try {
    const { userId } = req.params;
    const { limit = 10 } = req.query;
    
    const purchaseService = req.app.locals.services.purchase;
    
    if (!purchaseService) {
      return res.status(503).json({
        success: false,
        error: 'Purchase service not available'
      });
    }

    const purchases = await purchaseService.getUserPurchases(userId, parseInt(limit));
    
    res.json({
      success: true,
      data: purchases
    });
  } catch (error) {
    logger.error('Get user purchases error:', error);
    
    res.status(500).json({
      success: false,
      error: 'Failed to get user purchases'
    });
  }
});

module.exports = router;